repose-aggregator/components/filters/keystone-v2-filter/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Authorization.scala

```diff
@@ -71,83 +85,64 @@ object KeystoneV2Authorization extends LazyLogging {
         case _ =>
           logger.error("An unexpected tenant extraction type was encountered")
           None
-      }).find(_.nonEmpty).flatten
-    ).getOrElse(throw UnparsableTenantException("Could not parse tenant from the URI and/or the configured header"))
+      })
+    ).filter(_.nonEmpty).getOrElse(throw UnparsableTenantException("Could not parse tenant from the URI and/or the configured header"))
   }
 
-  def getTenantScopedRoles(config: ValidateTenantType, tenantToMatch: => String, roles: Seq[Role]): Seq[Role] = {
-    Option(config) match {
-      case Some(validateTenant) if !validateTenant.isEnableLegacyRolesMode =>
-        roles.filter(role =>
-          role.tenantId.forall(roleTenantId =>
-            roleTenantId.equals(tenantToMatch)))
-      case _ =>
-        roles
+  // NOTE: Replaces the tenant and roles scoping methods. For the roles scoping, also accounts for tenant prefixes, which was not previously the case.
+  def getScopedTenantToRolesMap(config: ValidateTenantType, tenantsToMatch: Set[String], tenantToRolesMap: TenantToRolesMap): TenantToRolesMap = {
+    val prefixes = Option(config.getStripTokenTenantPrefixes).map(_.split('/')).getOrElse(Array.empty[String])
+    tenantToRolesMap filterKeys { tenant =>
+      tenant == DomainRoleTenantKey || tenantsToMatch.exists(tenantToMatch =>
+        tenant == tenantToMatch || prefixes.exists(prefix =>
```

Comment:
we do this twice, extract to function?

Reply:
So... I thought about this about, but maybe you can come up with a better solution. The problem is, we are doing similar things, but not quite the same.

In one case, we want to scope the map down to just the relevant tenants and roles. So we have to iterate over all of the entries in the tenant-to-roles map to validate the tenant. We also want to keep prefixed tenants that match. Oh, and we also want to keep our special "tenant" for domain roles.

In the other case, we only need to verify that all of the tenants extracted from the request match something. We can't just do a size check between the request tenants and the scoped tenants because of prefixes, so I ended up iterating over the extracted tenants instead. There are two small optimizations here, though. First, we only pass the scoped tenants, so we know each user tenant matches some request tenant. Second, we stop as soon as any user tenant matches a request tenant and move on to the next request tenant.

I suppose I could write more code to track which request tenants matched when scoping the tenant-to-roles map, but then I conflate scoping and authorization, which I didn't want to do because it complicates things a little and the return type gets gross.

Or did you just mean extract the prefixed equality check? I can totally do that.

Reply:
Yeah, that's what i was talking about. the differences between line 97-98 and line 114-115 is a parenthesis at the end.

Reply:
Done.